projects/core/src/main/resources/data/computercraft/lua/rom/programs/shell.lua

```diff
@@ -54,6 +54,68 @@ else
     bgColour = colours.black
 end
 
+local function tokenise(...)
+    local sLine = table.concat({ ... }, " ")
+    local tWords = {}
+    local bQuoted = false
+    for match in string.gmatch(sLine .. "\"", "(.-)\"") do
+        if bQuoted then
+            table.insert(tWords, match)
+        else
+            for m in string.gmatch(match, "[^ \t]+") do
+                table.insert(tWords, m)
+            end
+        end
+        bQuoted = not bQuoted
+    end
+    return tWords
+end
+
+-- Execute a program using os.run, unless a shebang is present.
+-- In that case, execute the program using the interpreter specified in the shebang.
+-- This may occur recursively, up to the maximum number of times specified by nRemainingRecursion
+-- Returns the same type as os.run, which is a boolean indicating whether the program exited successfully.
+local function executeProgram(nRemainingRecursion, env, sPath, ...)
+    local file = fs.open(sPath, "r")
+    if not file then
+        printError("File not found: " .. sPath)
+        return false
+    end
+
+    -- First check if the file begins with a #!
+    local sContents = file.readLine()
+    if sContents:sub(1, 2) == "#!" then
+        nRemainingRecursion = nRemainingRecursion - 1
+        if nRemainingRecursion == 0 then
+            printError("Interpreter recursion depth limit reached when loading file: " .. sPath)
+            return false
+        end
+
+        -- Load the specified interpreter instead
+        local tParts = tokenise(sContents:sub(3))
+        if #tParts > 2 then
+            printError("Interpreter may only take 1 argument, but " .. #tParts - 1 .. " were specified: " .. table.concat(tParts, " "))
```

Comment:
Ah yeah, I did this to maintain compatibility with how it works in POSIX. But I suppose it's not really going to be compatible anyway since we won't have any of the standard program paths probably.